binr/ragg2/ragg2.c

```diff
@@ -108,6 +123,454 @@ static int openfile (const char *f, int x) {
+		}
+	}
+
+	if (!strcmp (os, "darwin")) {
+		if ((cEnv->OBJCOPY = strdup ("gobjcopy")) == NULL) {
+			eprintf ("Not enough memory for OBJCOPY allocation!\n");
+			goto fail;
+		}
+		if ((cEnv->FMT = strdup ("mach0")) == NULL) {
+			eprintf ("Not enough memory for FMT allocation!\n");
+			goto fail;
+		}
+		if (!strcmp (arch, "x86")) {
+			if (bits == 32) {
+				if ((cEnv->CFLAGS = strdup ("-arch i386")) == NULL) {
+					eprintf ("Not enough memory for CFLAGS!\n");
+					goto fail;
+				}
+				if ((cEnv->LDFLAGS = strdup ("-arch i386 -shared -c")) == NULL) {
+					eprintf ("Not enough memory for LDFLAGS!\n");
+					goto fail;
+				}
+			} else {
+				if ((cEnv->CFLAGS = strdup ("-arch x86_64")) == NULL) {
+					eprintf ("Not enough memory for CFLAGS!\n");
+					goto fail;
+				}
+				if ((cEnv->LDFLAGS = strdup ("-arch x86_64 -shared -c")) == NULL) {
+					eprintf ("Not enough memory for LDFLAGS!\n");
+					goto fail;
+				}
+			}
+		} else {
+			if ((cEnv->LDFLAGS = strdup ("-shared -c")) == NULL) {
+				eprintf ("Not enough memory for LDFLAGS!\n");
+				goto fail;
+			}
+		}
+		if ((cEnv->SHDR = r_str_newf ("\n.text\n%s _main\n", cEnv->JMP)) == NULL) {
+			eprintf ("Not enough memory for SHDR!\n");
+			goto fail;
+		}
+
+	} else {
+		if ((cEnv->OBJCOPY = strdup ("objcopy")) == NULL) {
+			eprintf ("Not enough memory for OBJCOPY allocation!\n");
+			goto fail;
+		}
+		if ((cEnv->FMT = strdup ("elf")) == NULL) {
+			eprintf ("Not enough space for FMT allocation!\n");
+			goto fail;
+		}
+		cEnv->SHDR = r_str_newf ("\n.section .text\n.globl  main\n"
+				   "// .type   main, @function\n%s main\n", cEnv->JMP);
+		if (cEnv->SHDR == NULL) {
+			eprintf ("Not enough memory for SHDR!\n");
+			goto fail;
+		}
+		if (!strcmp (arch, "x86")) {
+			if (bits == 32) {
+				cEnv->CFLAGS = strdup ("-fPIC -fPIE -pie -fpic -m32");
+				if (cEnv->CFLAGS == NULL) {
+					eprintf ("Not enough memory for CFLAGS");
+					goto fail;
+				}
+				cEnv->LDFLAGS = strdup ("-fPIC -fPIE -pie -fpic -m32");
+				if (cEnv->LDFLAGS == NULL) {
+					eprintf ("Not enough memory for LDFLAGS");
+					goto fail;
+				}
+			} else {
+				cEnv->CFLAGS = strdup ("-fPIC -fPIE -pie -fpic -m64");
+				if (cEnv->CFLAGS == NULL) {
+					eprintf ("Not enough memory for CFLAGS");
+					goto fail;
+				}
+				cEnv->LDFLAGS = strdup ("-fPIC -fPIE -pie -fpic -m64");
+				if (cEnv->LDFLAGS == NULL) {
+					eprintf ("Not enough memory for LDFLAGS");
+					goto fail;
+				}
+			}
+		} else {
+			cEnv->CFLAGS = strdup ("-fPIC -fPIE -pie -fpic -nostartfiles");
+			if (cEnv->CFLAGS == NULL) {
+				eprintf ("Not enough memory for CFLAGS");				
+				goto fail;
+			}
+
+			cEnv->LDFLAGS = strdup ("-fPIC -fPIE -pie -fpic -nostartfiles");
+			if (cEnv->LDFLAGS == NULL) {
+				eprintf ("Not enough memory for LDFLAGS");				
+				goto fail;
+			}
+		}
+	}
+
+	cEnv->TRIPLET = r_str_newf ("%s-%s-%d", os, arch, bits);
+
+	if (!strcmp (os, "windows")) {
+		if ((cEnv->TEXT = strdup (".text")) == NULL) {
```

Comment:
Use ! Instead of NULL but u dont need to chk nulptr here because u can check it once right after filling that env ina a check() function or so instead of checking eveything all the time